pkg/controller/gitsourceanalysis/gitsourceanalysis_controller.go

```diff
@@ -0,0 +1,155 @@
+	err := r.client.Get(context.TODO(), request.NamespacedName, gsAnalysis)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		reqLogger.Error(err, "There was an error while reading the GitSourceAnalysis object")
+		return reconcile.Result{}, err
+	}
+
+	if gsAnalysis.Status.Analyzed {
+		reqLogger.WithValues("git-source", gsAnalysis.Spec.GitSourceRef).
+			Info("Skipping GitSourceAnalysis as it was already analyzed")
+		return reconcile.Result{}, err
+	}
+
+	return analyze(reqLogger, r.client, gsAnalysis, request.Namespace)
+}
+
+func analyze(logger logr.Logger, client client.Client, gsAnalysis *v1alpha1.GitSourceAnalysis, namespace string) (reconcile.Result, error) {
+	gitSource := &v1alpha1.GitSource{}
+	err := client.Get(context.TODO(), newNsdName(namespace, gsAnalysis.Spec.GitSourceRef.Name), gitSource)
+	if err != nil {
+		gsAnalysis.Status.Error = "Failed to fetch the input source"
+		logger.WithValues("git-source", gsAnalysis.Spec.GitSourceRef).
+			Error(err, "There was an error while reading the GitSource object")
+
+	} else {
+		buildEnvStats, err := analyzeGitSource(log.LogWithGSValues(logger, gitSource), client, gitSource, namespace)
+		if err != nil {
+			gsAnalysis.Status.Error = err.Error()
+		} else {
+			gsAnalysis.Status.BuildEnvStatistics = *buildEnvStats
+		}
+	}
+
+	gsAnalysis.Status.Analyzed = true
+	err = client.Update(context.TODO(), gsAnalysis)
+	if err != nil {
+		log.LogWithGSValues(logger, gitSource).Error(err, "Error updating GitSourceAnalysis object")
+		if errors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
+			// Return and don't requeue
+			return reconcile.Result{}, nil
+		}
+		// Error updating the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func analyzeGitSource(logger *log.GitSourceLogger, client client.Client, gitSource *v1alpha1.GitSource, namespace string) (*v1alpha1.BuildEnvStats, error) {
+	logger.Info("Analyzing GitSource")
+
+	// Fetch the GitSource secret
+	gitSecretProvider, err := git.NewGitSecretProvider(client, namespace, gitSource)
+	if err != nil {
+		logger.WithValues("secret", gitSource.Spec.SecretRef.Name).
+			Error(err, "Error reading the secret object")
+		return nil, err
+
+	} else {
+		buildEnvStats, err := detector.DetectBuildEnvironments(logger, gitSource, gitSecretProvider)
+		if err != nil {
+			logger.Error(err, "Error detecting build types")
+		}
+		return buildEnvStats, err
+	}
+}
+
+func newNsdName(namespace, name string) types.NamespacedName {
```

Comment:
Could you please rename this to newNamespacedName(..) :D

Reply:
renamed ef3e10b